src/python/WMCore/MicroService/Tools/PycurlRucio.py

```diff
@@ -0,0 +1,267 @@
+    urls = ['{}/dids/{}/{}?dynamic=anything'.format(rucioUrl, scope, dset) for dset in datasets]
+    logging.info("Executing %d requests against Rucio for the container size", len(urls))
+    data = multi_getdata(urls, ckey(), cert(), headers=headers, decode=True)
+
+    for row in data:
+        dataset = row['url'].split('/dids/{}}/'.format(scope))[1]
+        dataset = dataset.replace("?dynamic=anything", "")
+        if row['data'] is None:
+            msg = "Failure in getPileupDatasetSizesRucio for dataset %s. Error: %s %s"
+            logging.info(msg, dataset, row.get('code'), row.get('error'))
+            sizeByDset.setdefault(dataset, None)
+            continue
+        sizeByDset.setdefault(dataset, row['data']['bytes'])
+    return sizeByDset
+
+
+def getPileupSubscriptionsRucio(datasets, rucioUrl, rucioToken, scope="cms"):
+    """
+    Provided a list of datasets, find dataset level subscriptions where it's
+    as complete as `percent_min`.
+    :param datasets: list of dataset names
+    :param rucioUrl: a string with the Rucio URL
+    :param rucioToken: a string with the user rucio token
+    :param scope: a string with the Rucio scope of our data
+    :return: a dictionary of datasets and a list of their location.
+    NOTE: Value `None` is returned in case the data-service failed to serve a given request.
+    """
+    # FIXME: we should definitely make a feature request to Rucio...
+    # so much, just to get the final RSEs for a dataset!!!
+    locationByDset = {}
+    if not datasets:
+        return locationByDset
+
+    headers = {"X-Rucio-Auth-Token": rucioToken}
+    # first, resolve the dataset into blocks
+    blocksByDset = getDatasetBlocksRucio(datasets, rucioUrl, rucioToken, scope)
+    urls = []
+    for _dset, blocks in blocksByDset.items():
+        if blocks:
+            for block in blocks:
+                urls.append('{}/replicas/{}/{}/datasets'.format(rucioUrl, scope, quote(block)))
+
+    # this is going to be bloody expensive in terms of HTTP requests
+    logging.info("Executing %d requests against Rucio replicas API for blocks", len(urls))
+    data = multi_getdata(urls, ckey(), cert(), headers=headers)
+    for row in data:
+        block = row['url'].split("/{}}/".format(scope))[1]
+        block = unquote(re.sub("/datasets$", "", block, 1))
+        dataset = block.split("#")[0]
+        locationByDset.setdefault(dataset, set())
+        if row['data'] is None:
+            msg = "Failure in getPileupSubscriptionsRucio for dataset {} and block {}".format(dataset, block)
+            msg += " Error: {} {}".format(row.get('code'), row.get('error'))
+            logging.info(msg)
+
+            locationByDset[dataset] = None
+            continue
+        if locationByDset.get(dataset, "non-None") is None:
+            # then one of the block requests failed, skip the whole dataset
+            continue
```

Comment:
Shouldn't we log our actions here too before we just continue?

Reply:
It will get logged at the `RequestInfo` level, but I agree that it's a good idea to log which query failed here. Log level a few lines above needs to be changed to error as well.

Reply:
Actually, this `if` will only be executed if the previous one has been evaluated to true (row[data] is None), for the same dataset. This means, one of the block level calls failed for this dataset, and in that case, we simply have to mark that dataset as "unresolved" for the moment and retry in the next cycle.